### ISSUE TITLE
`parse.d` static assert failure

### ISSUE THREAD

**Issue by jmh530** [libmir/mir-algorithm#282]
I am getting the following error running `dub test` with the master branch. I am on Windows 10 with DMD64 D Compiler v2.092.0-dirty. 

```
C:\Users\[removed]\Documents\GitHub\mir-algorithm>dub test
Generating test runner configuration 'mir-algorithm-test-default' for 'default' (library).
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Performing "unittest" build using dmd for x86_64.
mir-core 1.1.7: building configuration "library"...
mir-algorithm 3.9.0+commit.3.g7dff9ce: building configuration "mir-algorithm-test-default"...
Warning: multivariate cubic spline with derivatives was not tested!!!
Warning: multivariate cubic spline with derivatives was not tested!!!
source\mir\parse.d(89,9): Error: static assert:  fromString("3.518437208883201171875") == 3.51844L is false
dmd failed with exit code 1.
```

**Comment by 9il** [comment]
interesting, I thought DMD on windows hasn't 80 bit reals.

**Comment by jmh530** [comment]
@9il From the spec: "largest FP size implemented in hardware (Implementation Note: 80 bits for x86 CPUs or double size, whichever is larger)" Nothing about OS.

**Comment by 9il** [comment]
I am almost sure it is another DMD bug.

Could you please provide `fromString("3.518437208883201171875") ` and `3.518437208883201171875L` hexadecimal FP strings?

**Comment by jmh530** [comment]
I ran the code below
```
unittest {
    import std.stdio: writefln;
    import std.format;
    writefln("%1.21f", fromString!real("3.518437208883201171875"));
    writefln("%a", fromString!real("3.518437208883201171875"));
    writefln("%a", 3.518437208883201171875L);
}
```
and got 
```
3.518437208883200995047
0x1.c25c268497684p+1
0x1.c25c268497684p+1
```

**Comment by jmh530** [comment]
When I change that static assert to
`static assert("3.518437208883201171875".fromString!real == 0x1.c25c268497684p+1);`
it still fails with the same error message.

**Comment by 9il** [comment]
does `0xe.12e13424bb42p-2L` works? With L at the end?

**Comment by 9il** [comment]
Also, please provide more information. Is the test passes with 2.089 for example? What target is used? Does it use MSVC? is 32 bit or 64 bit?

D spec is bad in terms of FP and isn't really represent the state of implementation. Also, D according to the spec isn't IEEE compatible language, so it can't be used for math. At the same time in practice, we can write a code that is IEEE compatible because fortunately, another part of the world including LLVM and GCC doesn't think 80-bit reals are good to handle single/double precision math.

**Comment by jmh530** [comment]
I tried what you suggested above and it doesn't work. 

On more information, I will not be able to test old compilers until tomorrow or the day after as I usually do not keep multiple versions installed or use dvm. It compiles for x86_64 by default. I also tried `dub arch --arch=x86` and had the same issue.  I also had the same issue with ldc2 1.21.0. On MSVC, I am running it from the command line, just navigating the folder and running `dub test`.  The x86_64 build on Windows now requires MSVC to be installed to use the linker, however. I believe --arch=x86, though still uses optlink. I don't think I am doing anything special with targets...

I tried `writefln("%a", fromString!real("3.518437208883201171875") - 0xe.12e13424bb42p-2L);` and the result was `0x1.978p-53`. The following also fails `static assert(3.518437208883201171875L == 0xe.12e13424bb42p-2);`. I adjusted one of the other writefln's with `writefln("%1.57f", fromString!real("3.518437208883201171875"));` and got `3.518437208883200995046536263544112443923950195312500000000`

**Comment by 9il** [comment]
#286 updates LDC version for Windows CI. All works well.
MSVC `real` type is the same as `double`, the failing test should never be run.

Posix cross-compilation gives the same answer.

```d
static real d = -12.3e-30L;
static double e = -12.3e-30L;
static assert(real.mant_dig == 53);
```

when compiled with 
```
-mtriple=i686-windows-msvc
```

gives

```s
real example.d:
        .quad   -5048764808226925355

double example.e:
        .quad   -5048764808226925355
```

**Comment by 9il** [comment]
https://forum.dlang.org/post/wyhpritoexkatdbvpfde@forum.dlang.org

Please note, that this lines can fail on Windows and this isn't the bug. The bug here that this lines are compiled for Windows while they are guarded with `static if (real.mant_dig >= 64)`.

**Comment by 9il** [comment]
@jmh530 Please append `L` for hexadecimal results.
It should be 
```
static assert(3.518437208883201171875L == 0xe.12e13424bb42p-2L);
```
and
```
static assert(fromString!real("3.518437208883201171875") == 0xe.12e13424bb42p-2L);
```

The hexadecimal FP literals in both asserts have trailing `L`.

What asserts are failing and what asserts pass?

**Comment by jmh530** [comment]
@9il With `dub test --arch=x86_64`, both fail.
With  `dub test --arch=x86`, both fail.
With `dub test --arch=x86_64-pc-windows-msvc --compiler=ldc2`, the second one passes. 
With `dub test --arch=x86_64-pc-windows-mingw --compiler=ldc2`, the second one passes. 
With `dub test --arch=i686-pc-windows-msvc --compiler=ldc2`, gives a linker error (as does the mingw one). 

After reading what `kinke` said on that thread, I was surprised to find that when compiling with `dub test --arch=x86_64-pc-windows-mingw --compiler=ldc2`, the `static if (real.mant_dig >= 64)` does not compile the code within it. `dub test --arch=x86_64` will compile it. 

I will not be available again until tomorrow.

**Comment by 9il** [comment]
please check if it fixed now

**Comment by jmh530** [comment]
It seems as if you have fixed the static assert and one assert failure, but there is still another assert failure

I just tested this on my laptop with 2.092.0-dirty and did not get the static assert failure, but did get the assert failure with the message below. 

```
C:\Users\[removed]\Documents\GitHub\mir-algorithm>dub test --arch=x86_64 --compiler=dmd
Generating test runner configuration 'mir-algorithm-test-default' for 'default' (library).
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
Performing "unittest" build using dmd for x86_64.
mir-core 1.1.7: building configuration "library"...
mir-algorithm 3.9.0+commit.7.g45ecbbc: building configuration "mir-algorithm-test-default"...
Warning: multivariate cubic spline with derivatives was not tested!!!
Warning: multivariate cubic spline with derivatives was not tested!!!
Linking...
   Creating library .dub\build\mir-algorithm-test-default-unittest-windows-x86_64-dmd_2092-78B722E466FC0CF0898EC5C3E7AC97CF\mir-algorithm-test-default.lib and object .dub\build\mir-algorithm-test-default-unittest-windows-x86_64-dmd_2092-78B722E466FC0CF0898EC5C3E7AC97CF\mir-algorithm-test-default.exp
Running .\mir-algorithm-test-default.exe
source\mir\bignum\low_level_view.d(1925): [unittest] unittest failure

core.exception.AssertError@source\mir\bignum\low_level_view.d(1925): unittest failure
----------------
0x00007FF6E512D632 in d_unittestp
0x00007FF6E4A49A06 in mir.bignum.low_level_view.__unittest_L1850_C1 at C:\Users\[removed]\Documents\GitHub\mir-algorithm\source\mir\bignum\low_level_view.d(1925)
0x00007FF6E4ABA01E in void mir.bignum.low_level_view.__modtest()
0x00007FF6E5155EE9 in int core.runtime.runModuleUnitTests().__foreachbody1(object.ModuleInfo*)
0x00007FF6E51455F7 in int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*))
0x00007FF6E515CB44 in int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_win64.SectionGroup)
0x00007FF6E515CAAF in int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*)))
0x00007FF6E51455B7 in int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*))
0x00007FF6E5155D1E in runModuleUnitTests
0x00007FF6E513FCA5 in void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll()
0x00007FF6E513FC0C in void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).tryExec(scope void delegate())
0x00007FF6E513FA43 in d_run_main2
0x00007FF6E5123990 in d_run_main
0x00007FF6E4869782 in dub_test_root._d_cmain!().main at C:\D\dmd2\windows\bin64\..\..\src\druntime\import\core\internal\entrypoint.d(29)
0x00007FF6E5217AC4 in __scrt_common_main_seh at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl(288)
0x00007FFD2D9A6FD4 in BaseThreadInitThunk
0x00007FFD2DADCEC1 in RtlUserThreadStart
1/54 unittests FAILED
Program exited with code 1
```

I noticed that my desktop and laptop are using different versions of DMD. The error report came from using DMD64 on the laptop, while my desktop was set up to DMD32. I apparently had adjusted my laptop settings to use DMD64, which just become available in [2.091](https://dlang.org/changelog/2.091.0.html#windows), but not the desktop. Specifically, the desktop computer is running DMD 2.092.1-dirty. 

With the desktop using DMD32 and trying to avoid ambiguity, I tried `dub test --arch=x86_64 --compiler=dmd` and I got `dmd failed with exit code -1073741819.`. When I tried `dub test --arch=x86 --compiler=dmd` I got 
```
Error: more than 32767 symbols in object file
dmd failed with exit code 1.
```

With the desktop using DMD64 and trying to avoid any ambiguity, I tried `dub test --arch=x86_64 --compiler=dmd` and I got essentially the same assert up at the top. When I tried `dub test --arch=x86 --compiler=dmd` I got 
```
Error: more than 32767 symbols in object file
dmd failed with exit code 1.
```

**Comment by 9il** [comment]
Thank you a lot!

DMD bugs filled.

 - [wrong real literals on windows](https://issues.dlang.org/show_bug.cgi?id=20972)
 - [Compiler error: more than 32767 symbols in object file](https://issues.dlang.org/show_bug.cgi?id=20973)
